### ISSUE TITLE
[Alpha5] Can't install any package. Bad error message about syntax error (which location is not obvious)

### ISSUE THREAD

**Issue by Kakadu** [ocaml/opam#2708]
It complains during installation of ocp-indent but `opam` file for ocp-indent itself should be OK. Actually, I can't install any package now....

```
➜  opam git:(36d3112) ✗ opam install ocp-indent --debug
00:00.002  GSTATE                  LOAD-GLOBAL-STATE
00:00.002  RSTATE                  LOAD-REPOSITORY-STATE
00:00.025  RSTATE                  Loaded /home/kakadu/.opam/repo/state.cache in 0.000s
00:00.038  RSTATE                  Cache found
00:00.038  STATE                   LOAD-SWITCH-STATE
00:00.057  STATE                   Detected changed packages (marked for reinstall): {}
00:00.057  STATE                   Switch state loaded in 0.000s
00:00.057  FILE(switch-state)      Wrote /home/kakadu/.opam/4.03.0/.opam-switch/backup/state-20161007093821.export in 0.000s
00:00.074  CLIENT                  INSTALL ocp-indent
00:00.131  SYSTEM                  mkdir /home/kakadu/.opam/log
00:00.137  SYSTEM                  [log-32283-4438d8] (in 0.000s) ocamlc -vnum
00:00.182  CLIENT                  Orphans: full {}, versions {}
00:00.268  SOLVER                  resolve request=install:(ocp-indent) remove:() upgrade:()
00:00.278  SYSTEM                  rm /home/kakadu/.opam/4.03.0/.opam-switch/backup/state-20161007093821.export
Fatal error: Invalid character   in package version "0.20.0 & <= 0.21.0"
00:00.278  SYSTEM                  rmdir /home/kakadu/.opam/log
00:00.279  SYSTEM                  rmdir /tmp/opam-kakadu-32283
➜  opam git:(36d3112) ✗ opam show ocp-indent      

=-=- ocp-indent: information on all versions =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
name         ocp-indent
all-versions 0.1.0  0.6.0  0.6.1  0.6.2  0.9.0  0.9.2  1.0.0  1.0.1  1.0.2  1.1.0  1.2.0  1.2.1  1.2.2  1.3.0  1.3.1  1.3.2  1.4.0  1.4.1  1.4.2  1.5  1.5.1  1.5.2  1.5.3

=-=- Version-specific details -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
version       1.5.3
repository    default
url.src:      https://github.com/OCamlPro/ocp-indent/archive/1.5.3.tar.gz
url.checksum: md5=7515340b23df8884e4ae6f6a4315d5f7
homepage:     http://www.typerex.org/ocp-indent.html
bug-reports:  https://github.com/OCamlPro/ocp-indent/issues
dev-repo:     git+https://github.com/OCamlPro/ocp-indent.git
authors:      "Louis Gesbert <louis.gesbert@ocamlpro.com>"
              "Thomas Gazagnaire <thomas@gazagnaire.org>"
              "Jun Furuse"
license:      LGPL
tags:         "org:ocamlpro" "org:typerex"
flags:        light-uninstall
depends:      "ocaml" "ocp-build" {>= "1.99.6-beta"} "cmdliner"
synopsis      A simple tool to indent OCaml programs
description   Ocp-indent is based on an approximate, tolerant OCaml parser and a simple stack
              machine ; this is much faster and more reliable than using regexps. Presets and
              configuration options available, with the possibility to set them project-wide.
              Supports most common syntax extensions, and extensible for others.
              
              Includes:
              
              * An indentor program, callable from the command-line or from within editors
              * Bindings for popular editors
              * A library that can be directly used by editor writers, or just for
              approximate parsing.
➜  opam git:(36d3112) ✗ opam show ocp-indent --raw
opam-version: "2.0"
name: "ocp-indent"
version: "1.5.3"
synopsis: "A simple tool to indent OCaml programs"
description: """
Ocp-indent is based on an approximate, tolerant OCaml parser and a simple stack
machine ; this is much faster and more reliable than using regexps. Presets and
configuration options available, with the possibility to set them project-wide.
Supports most common syntax extensions, and extensible for others.

Includes:

* An indentor program, callable from the command-line or from within editors
* Bindings for popular editors
* A library that can be directly used by editor writers, or just for
approximate parsing."""
maintainer: "contact@ocamlpro.com"
authors: [
  "Louis Gesbert <louis.gesbert@ocamlpro.com>"
  "Thomas Gazagnaire <thomas@gazagnaire.org>"
  "Jun Furuse"
]
license: "LGPL"
tags: ["org:ocamlpro" "org:typerex"]
homepage: "http://www.typerex.org/ocp-indent.html"
bug-reports: "https://github.com/OCamlPro/ocp-indent/issues"
depends: [
  "ocaml"
  "ocp-build" {>= "1.99.6-beta"}
  "cmdliner"
]
flags: light-uninstall
build: [
  ["./configure" "--prefix" prefix]
  [make]
]
post-messages:
  """
This package requires additional configuration for use in editors. Install package 'user-setup', or manually:

* for Emacs, add these lines to ~/.emacs:
  (add-to-list 'load-path "%{share}%/emacs/site-lisp")
  (require 'ocp-indent)

* for Vim, add this line to ~/.vimrc:
  set rtp^="%{share}%/ocp-indent/vim\""""
    {success & !user-setup:installed}
dev-repo: "git+https://github.com/OCamlPro/ocp-indent.git"
url {
  src: "https://github.com/OCamlPro/ocp-indent/archive/1.5.3.tar.gz"
  checksum: "md5=7515340b23df8884e4ae6f6a4315d5f7"
}
➜  opam git:(36d3112) ✗ opam --version            
2.0~alpha5
```

**Comment by Kakadu** [comment]
```
➜  bap-server git:(master) ✗ grep "0.20.0 & <= 0.21.0" ~/.opam/repo/default -r                             
/home/kakadu/.opam/repo/default/packages/bap-server/bap-server.0.1.0/opam:29:  "cohttp" {>= "0.20.0 & <= 0.21.0"}
```

**Comment by Kakadu** [comment]
I tried to repin it with right  opam file (that passes `opam lint`) but this doesn't help.

**Comment by Kakadu** [comment]
I tried to remove this recipe from ~/.opam and after removing `state.cache` file it works without error.

**Comment by Kakadu** [comment]
So, the right title for the issue is that: _We should ignore opam recipes with errors_

**Comment by AltGr** [comment]
Thanks for reporting! And sorry for the trouble...
See https://github.com/ocaml/opam-repository/commit/da1962e3bc86f7ce14ae4e73aa7ebc749bb7adb1

The error is that, while recently making the rules for parsing versions stricter, I forgot to add a try/catch at the right spot... The expected behaviour is indeed to ignore errors in opam files (although we should keep a way to notify about them, just when necessary, which is being discussed in #2703)

Pinning didn't help in this case, because while the pinned-to opam file overrides the repository file, it doesn't prevent it to be read (the repository cache of loaded opam files being shared by all switches, it can't depend on what is pineed).

Expect a patch for the actual bug shortly.

**Comment by AltGr** [comment]
> We should ignore opam recipes with errors

it's what was done, but that specific error somehow escaped (since version constraints are now filtered, it was raised during the resolution of the filter, instead of during parsing). Thanks for reporting, should be fixed by #2712
